src/main/java/ch/njol/skript/config/Config.java

```diff
@@ -30,285 +15,367 @@
 	}
-	
+
+	/**
+	 * @deprecated This copies all values from the other config and sets them in this config,
+	 * which could be destructive for sensitive data if something goes wrong.
+	 * Also removes user comments.
+	 * Use {@link #updateNodes(Config)} instead.
+	 */
+	@Deprecated(forRemoval = true)
 	public boolean setValues(final Config other, final String... excluded) {
 		return getMainNode().setValues(other.getMainNode(), excluded);
 	}
 
 	/**
-	 * Compares the keys and values of this Config and another.
-	 * @param other The other Config.
-	 * @param excluded Keys to exclude from this comparison.
-	 * @return True if there are differences in the keys and their values
-	 *  of this Config and the other Config.
+	 * Updates the nodes of this config with the nodes of another config.
+	 * Used for updating a config file to a newer version.
+	 * <p>
+	 * This method only sets nodes that are missing in this config, thus preserving any existing values.
+	 * </p>
+	 *
+	 * @param newer The newer config to update from.
+	 * @return True if any keys were added to this config, false otherwise.
 	 */
-	public boolean compareValues(Config other, String... excluded) {
-		return getMainNode().compareValues(other.getMainNode(), excluded);
-	}
-	
-	@Nullable
-	public File getFile() {
-		if (file != null) {
-			try {
-				return file.toFile();
-			} catch (Exception e) {
-				return null; // ZipPath, for example, throws undocumented exception
+	public boolean updateNodes(@NotNull Config newer) {
+		Set<Node> newNodes = findNodes(newer.getMainNode());
+		Set<Node> oldNodes = findNodes(getMainNode());
+
+		newNodes.removeAll(oldNodes);
+		Set<Node> nodesToUpdate = new LinkedHashSet<>(newNodes);
+
+		if (nodesToUpdate.isEmpty())
+			return false;
+
+		for (Node node : nodesToUpdate) {
+			Skript.debug("Updating node %s", node);
+			SectionNode newParent = node.getParent();
+			Preconditions.checkNotNull(newParent);
+
+			SectionNode parent = getNode(newParent.getPath());
+			Preconditions.checkNotNull(parent);
+
+			int idx = node.getIndex();
+			if (idx >= parent.size()) {
+				Skript.debug("Adding node %s to %s (size mismatch)", node, parent);
+				parent.add(node);
+				continue;
+			}
+
+			Node existing = parent.getAt(idx);
+			if (existing != null) { // insert between existing
+				Skript.debug("Adding node %s to %s at index %s", node, parent, idx);
+				parent.add(idx, node);
+			} else {
+				Skript.debug("Adding node %s to %s", node, parent);
+				parent.add(node);
 			}
 		}
-		return null;
+		return true;
 	}
-	
-	@Nullable
-	public Path getPath() {
-		return file;
+
+	/**
+	 * Recursively finds <i>all</i> nodes in a section node, including other
+	 * section nodes, entry nodes, and void nodes.
+	 *
+	 * @param node The parent node to search.
+	 * @return A set of the discovered nodes, guaranteed to be in the order of discovery.
+	 */
+	@Contract(pure = true)
+	private static @NotNull Set<Node> findNodes(@NotNull SectionNode node) {
```

Comment:
I think this should be discoverNodes, since find implies we're trying to find something specific. Or flattenNodes()